File: LeetCode/이진탐색 3 제곱의 합.py
class Solution:
    def judgeSquareSum(self, c: int) -> bool:
        l = 0
        r = (int)(sqrt(c))
        while l <= r:
            m = l*l+r*r
            if m == c:
                return True
            elif m > c:
                r -= 1
            else:
                l += 1
        return False

File: LeetCode/이진탐색 3 제곱의 합.cpp
class Solution {
public:
    bool judgeSquareSum(int c) {
        // 상한값은 c의 제곱근
        int l = 0, r = sqrt(c);
        while (l <= r) {
            // 내부에서 가상의 합을 구해 c와 비교
            // 더 크면 r를 감소, 더 작으면 l을 증가
            long long m = (long long)(l*l)+(long long)(r*r);
            if (m == c) {
                return true;
            } else if (m > c) {
                r--;
            } else {
                l++;
            }
        }
        return false;
    }
};